com.unity.shadergraph/Editor/Generation/Processors/Generator.cs

```diff
@@ -190,53 +250,93 @@ void BuildShader()
+                var shaderCustomEditors = new List<ShaderCustomEditor>();
+                string shaderCustomEditor = typeof(GenericShaderGraphMaterialGUI).FullName;
+                string shaderFallback = "Hidden/Shader Graph/FallbackError";
+
                 GenerationUtils.GeneratePropertiesBlock(m_Builder, shaderProperties, shaderKeywords, m_Mode, graphInputOrderData);
-                for (int i = 0; i < m_Targets.Length; i++)
+                for (int i = 0; i < m_Targets.Count; i++)
                 {
-                    TargetSetupContext context = new TargetSetupContext(m_assetCollection);
+                    var context = m_TargetContexts[i];
 
-                    // Instead of setup target, we can also just do get context
-                    m_Targets[i].Setup(ref context);
-
-                    var subShaderProperties = GetSubShaderPropertiesForTarget(m_Targets[i], m_GraphData, m_Mode, m_OutputNode, m_TemporaryBlocks);
-                    foreach (var subShader in context.subShaders)
+                    // process the subshaders
+                    var subShaderProperties = GetSubShaderPropertiesForTarget(m_Targets[i], m_GraphData, m_Mode, m_OutputNode, isPrimaryShader ? m_PrimaryShaderTemporaryBlocks : null);
+                    foreach (SubShaderDescriptor subShader in context.subShaders)
                     {
-                        GenerateSubShader(i, subShader, subShaderProperties);
-                    }
+                        // only generate subshaders that belong to the current shader we are building
+                        if (subShader.additionalShaderID == additionalShaderID)
+                        {
+                            GenerateSubShader(i, subShader, subShaderProperties);
 
-                    var customEditor = context.defaultShaderGUI;
-                    if (customEditor != null && m_Targets[i].WorksWithSRP(GraphicsSettings.currentRenderPipeline))
-                    {
-                        m_Builder.AppendLine("CustomEditor \"" + customEditor + "\"");
-                    }
+                            // pull out shader data from the subshader
+                            if (subShader.shaderDependencies != null)
+                                shaderDependencies.AddRange(subShader.shaderDependencies);
 
-                    foreach (var rpCustomEditor in context.customEditorForRenderPipelines)
-                    {
-                        m_Builder.AppendLine($"CustomEditorForRenderPipeline \"{rpCustomEditor.shaderGUI}\" \"{rpCustomEditor.renderPipelineAssetType}\"");
+                            if (subShader.shaderCustomEditor != null)
+                                shaderCustomEditor = subShader.shaderCustomEditor;
+
+                            if (subShader.shaderCustomEditors != null)
+                                shaderCustomEditors.AddRange(subShader.shaderCustomEditors);
+
+                            if (subShader.shaderFallback != null)
```

Comment:
Similar question about custom editor. Does the last win and is that expected?

Reply:
yes.  Other option is to make it a conflict error or warning if we get multiple non-matching specifications for these....